docs/testing.rst

```diff
@@ -0,0 +1,98 @@
+=======
+Testing best practices
+=======
+
+Test set up
+===========
+
+Doing a lot of work in the ``setUp`` call of a test class means that it will be run on every test. This
+quickly adds a lot of run time to the tests. Some things that can be easily moved to ``setUpClass`` are domain
+creation, user creation, or any other static models needed for the test.
+
+Furthermore, nose provides the ability to write setup for the entire package and/or module. More information
+on that can be found `here <http://pythontesting.net/framework/nose/nose-fixture-reference/#package>`_.
+
+Using SimpleTestCase
+====================
+
+The SimpleTestCase runs tests without a database. Many times this can be achieved through the use of the `mock
+library <http://www.voidspace.org.uk/python/mock/>`_. A good rule of thumb is to have 80% of your tests be unit
+tests that utilize ``SimpleTestCase``, and then 20% of your tests be integration tests that utilize the
+database and ``TestCase``.
+
+CommCareHQ also has some custom in mocking tools.
+
+    - `Fake Couch <https://github.com/dimagi/fakecouch>`_ - Fake implementation of CouchDBKit api for testing purposes.
+    - `ESQueryFake <https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/es/fake/es_query_fake.py>`_ - For faking ES queries.
+
+
+Tearing down tests
+==================
+
+It is important to ensure that all objects you have created in the test database are deleted when the test
+class finishes running. This often happens in the ``tearDown`` method or the ``tearDownClass`` method.
+However, unneccessary cleanup "just to be safe" can add a large amount of time onto your tests.
+
+Combining test classes
+======================
+
+Sometimes classes share the same base class and inherit the ``setUpClass`` function. Below is an example:
+
+.. code:: python
+    # BETTER EXAMPLE
+
+    class MyTextMixin(object)
+
+        def _my_helper_function(self):
+            ...
+
+    class MyTestClass(TestCase, MyTextMixin):
+
+        @classmethod
+        def setUpClass(cls):
+            ...
+
+        def test(self):
+            ...
+
+    class MyTestClassTwo(TestCase, MyTextMixin):
+
+        @classmethod
+        def setUpClass(cls):
+            ...
+
+        def test(self):
+            ...
+
+.. code:: python
+    # BAD EXAMPLE
+
+    class MyBaseTestClass(TestCase):
+
+        @classmethod
+        def setUpClass(cls):
+            ...
+
+        def _my_helper_function(self):
+            ...
+
+
+    class MyTestClass(MyBaseTestClass):
+
+        def test(self):
+            ...
+
+    class MyTestClassTwo(MyBaseTestClass):
+
+        def test(self):
+            ...
+
+
+In the above example the ``setUpClass`` is run three times, once for ``MyBaseTestClass``, ``MyTestClass`` and ``MyTestClassTwo``. If
```

Comment:
> ``setUpClass`` is run three times

In `BAD EXAMPLE`, this is not true, it's only run twice because `MyBaseTestClass` has no test methods.

Also, you say "In the above example" but both GOOD and BAD examples are above and it's not clear which you're referring to, and the GOOD example is above the BAD example, so that adds to the confusion.